otter/metrics.py

```diff
@@ -253,6 +254,40 @@ def connect_cass_servers(reactor, config):
         seed_endpoints, config['keyspace'], disconnect_on_cancel=True)
 
 
+def log_divergent_groups(clock, divergent_groups, log, timeout, group_metrics):
+    """
+    Log groups that have not changed and been divergent for long time
+    """
+    dg = divergent_groups
+    converged, diverged = partition_bool(
+        lambda gm: gm.actual + gm.pending == gm.desired, group_metrics)
+    # stop tracking all converged groups
+    for gm in converged:
+        dg.pop((gm.tenant_id, gm.group_id), None)
+    # Start tracking divergent groups depending on whether they've changed
+    now = clock.seconds()
+    for gm in diverged:
+        pair = (gm.tenant_id, gm.group_id)
+        if pair in dg:
+            last_time, values = dg[pair]
+            if values != hash((gm.desired, gm.actual, gm.pending)):
+                del dg[pair]
```

Comment:
Sorry, to be clear, I'm not saying this change should be made, I just wanted to check with you that this isn't an issue (I don't think it is).

But let's say metrics runs every **m** seconds, and the timeout is **n** seconds.

At time `T=0`, group 1 is divergent with hash `a`, so it's put in the divergent list.
At time `T=m`, group 1 is divergent but with a different hash `b`, so it's removed from the divergent list.
At time `T=2m`, group 1 is still divergent with hash `b`, so it is put back in the divergent queue.
At time `T=2m + n`, group 1 is logged as divergent for `n` seconds, but actually it's been divergent for `n + m` seconds, since it's has hash `b` since time `T=m`, not `T=2m`.